File: Binary_Search/033.Search-In-Rotated-Sorted-Array/033.Search-In-Rotated-Sorted-Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low = 0
        high = len(nums) - 1
        # find the rotation point
        while low < high:
            middle = (low + high) // 2
            if nums[middle] < nums[0]:
                high = middle
            else:
                low = middle + 1
        pivot = high
        if target >= nums[0]:
            return self.bs(nums, 0, pivot, target)
        else:
            return self.bs(nums, pivot, len(nums) - 1, target)

    def bs(self, nums, low, high, target):
        while low <= high:
            middle = (low + high) // 2
            if nums[middle] > target:
                high = middle - 1
            elif nums[middle] < target:
                low = middle + 1
            else:
                return middle
        return -1

File: Binary_Search/033.Search-In-Rotated-Sorted-Array/033.Search-In-Rotated-Sorted-Array.java
class Solution {
    public int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        while (low < high) {
            int middle = (low + high) / 2;
            if (nums[middle] < nums[0]) {
                high = middle;
            } else {
                low = middle + 1;
            }
        }
        int pivot = high;
        if (target >= nums[0]) {
            return bs(nums, 0, pivot, target);
        } else {
            return bs(nums, pivot, nums.length - 1, target);
        }
    }

    public int bs(int[] nums, int low, int high, int target) {
        while (low <= high) {
            int middle = (low + high) / 2;
            if (nums[middle] > target) {
                high = middle - 1;
            } else if (nums[middle] < target) {
                low = middle + 1;
            } else {
                return middle;
            }
        }
        return -1;
    }
}
